### DISCUSSION TITLE
How does `pnpm-lock.yaml` determine what shasum is applied?

### DISCUSSION THREAD

**Discussion by Aghassi** [pnpm/pnpm#6194] [Q&A] [answered]
We have two devs at my company that both have run `pnpm i` on changed package.json. Mine resolves with `sha512`, while theirs resolves with `sha1`. How is this determined? Can we hardcode it to always be one or the other?

The only thing I've seen related to this sort topic is https://github.com/pnpm/pnpm/discussions/6070

Example diff

```diff
git diff pnpm-lock.yaml                                                        ✔
diff --git a/pnpm-lock.yaml b/pnpm-lock.yaml
index 2ee40b4b6a4b..3f15b0faf144 100644
--- a/pnpm-lock.yaml
+++ b/pnpm-lock.yaml
@@ -26,7 +26,7 @@ importers:
       '@babel/template': 7.18.6
       '@babel/traverse': 7.18.6
       '@babel/types': 7.18.7
-      '@emotion/babel-preset-css-prop': 10.1.0
+      '@emotion/babel-preset-css-prop': 11.10.0
       '@robinhood/babel-preset': workspace:*
       '@robinhood/caf-ts': workspace:*
       '@robinhood/crypto-yields-protobufjs': workspace:*
@@ -144,7 +144,7 @@ importers:
       '@babel/template': 7.18.6
       '@babel/traverse': 7.18.6
       '@babel/types': 7.18.7
-      '@emotion/babel-preset-css-prop': 10.1.0_@babel+core@7.18.6
+      '@emotion/babel-preset-css-prop': 11.10.0_@babel+core@7.18.6
       '@swc/core': 1.3.36
       '@typescript-eslint/eslint-plugin': 5.51.0_kjzbcm3fu7arvi5o6d6ax6aw2i
       '@typescript-eslint/parser': 5.51.0_cl6qtdo3npn4kngw26aibg642e
@@ -2048,8 +2048,8 @@ packages:
     engines: {node: '>=10.0.0'}
     dev: true

-  /@emotion/babel-plugin-jsx-pragmatic/0.1.5_@babel+core@7.18.6:
-    resolution: {integrity: sha512-y+3AJ0SItMDaAgGPVkQBC/S/BaqaPACkQ6MyCI2CUlrjTxKttTVfD3TMtcs7vLEcLxqzZ1xiG0vzwCXjhopawQ==}
+  /@emotion/babel-plugin-jsx-pragmatic/0.2.0_@babel+core@7.18.6:
+    resolution: {integrity: sha1-b914YAQXlz+iYQcEaTFYGB2FBbc=}
     peerDependencies:
       '@babel/core': ^7.0.0
     dependencies:
@@ -2057,43 +2057,58 @@ packages:
       '@babel/plugin-syntax-jsx': 7.18.6_@babel+core@7.18.6
     dev: true

-  /@emotion/babel-preset-css-prop/10.1.0_@babel+core@7.18.6:
-    resolution: {integrity: sha512-t1ar2Zt3fJ/TuoEg7Oin4sYdYt4BMWbvsQkaO4rq0II4hb9t/NdbCUd0jXFIYDuf0FYhyDEHt6sqpdPAG4uQrA==}
+  /@emotion/babel-plugin/11.10.6:
+    resolution: {integrity: sha1-po7ksBnWYdbzfexLiQMlV2aSXq0=}
```

**Comment by zkochan** [comment, marked as answer]
sha1 is only used if sha512 is not present in the package metadata. If you use the standard registry (registry.npmjs.org), then I think it should return sha512 for every package in the registry. They used to return sha1 in the past. You can try to run `pnpm store prune` to remove the local metadata cache but unlikely it is related to cache because I remember that npm upgraded the integrities months ago.

**Reply by Aghassi** [reply]
Ah ok in this case I think it's because we recently fixed our rc to point to our internal endpoint, and that is on Artifactory which has max support of SHA256. Thanks for the info! That makes sense.

**Comment by gluxon** [comment]
I've seen this as well. My conclusion was that there's a (still) ongoing database migration on `registry.npmjs.org` adding `dist.integrity` fields to existing packages. Packages published before the [npm registry supported sha512 in 2017](https://github.com/npm/npm-registry-client/pull/157) are getting the new `dist.integrity` field retroactively added. It wasn't a "every package gets this new field instantly" switch. The migration is happening incrementally, and affects both pnpm and [npm](https://stackoverflow.com/questions/47638381/why-did-package-lock-json-change-the-integrity-hash-from-sha1-to-sha512).

At this point most packages on the npm registry should have the new field, but every once in a while I do notice a very old package switch over. Developers on my team (and my own machine) can sometimes cache an older version of the local package metadata that does not contain `dist.integrity`. (As zkochan mentioned.) We noticed most of the flip-flopping stop after asking everyone to clear their local metadata cache.
